### PR TITLE
[ENH] Infer more things in Scan.infer_shape()

### PR DESCRIPTION
**Pull request by carriepl**

### CODE REVIEW

File: theano/scan_module/tests/test_scan.py [hunk 1]
```diff
@@ -3967,8 +3971,16 @@ def onestep(x, x_tm4):
...
         f = theano.function([seq], results[1], updates=updates)
+
+        # This fails if scan uses wrongly the borrow flag
         assert numpy.all(exp_out == f(inp))
 
+        # This fails if Scan's infer_shape() is unable to remove the Scan
+        # node from the graph.
+        f_infershape = theano.function([seq], results[1].shape)
+        scan_nodes_infershape = scan_nodes_from_fct(f_infershape)
+        assert(len(scan_nodes_infershape) == 0)
```
Comment (nouiz) [inline theano/scan_module/tests/test_scan.py]: This fail in FAST_COMPIILE. I think we should enable the local_shape_to_shape_i in FAST_COMPILE. This is fast to execute, the feature is already registered and this can remove much computation.

in tensor/opt.py, register local_shape_to_shape_i  like  `@register_canonicalize('fast_compile')

Comment (nouiz) [inline theano/scan_module/tests/test_scan.py]: This will fix the travis error, they are in FAST_COMPILE

**Comment by nouiz** [comment]
I'm ok with this, but it would be good if @lamblin or @abergeron also check this PR.

Also, maybe we are better to merge this Monday then today.

**Comment by carriepl** [comment]
It seems registering local_shape_to_shape_i in FAST_COMPILE makes a bunch of tests fail. So either
a) we don't keep it in FAST_COMPILE
b) we keep it in FAST_COMPILE but I adapt the tests.
As @nouiz, this depends on what @lamblin says about registering it in FAST_COMPILE in the first place.

File: theano/scan_module/tests/test_scan.py (hunk 1 above)
Comment (lamblin) [inline theano/scan_module/tests/test_scan.py]: Is the ShapeFeature registered in FAST_COMPILE? If not, adding it could slow down the whole compilation phase. We should profile that.
I don't think we expect shape inference to be performed in FAST_COMPILE in general, right? So we could just force the mode here for the tests.

Comment (lamblin) [inline theano/scan_module/tests/test_scan.py]: In fact, in `InferShapeTester`, we [add 'canonicalize'](https://github.com/Theano/Theano/blob/master/theano/tests/unittest_tools.py#L184) because of that.
If we want to include `local_shape_to_shape_i` in `'fast_compile'`, that could be reasonable, but it seems to be an orthogonal change, and we should also change `InferShapeTester` to be consistent.

**Comment by carriepl** [comment]
I forced the new test to use FAST_RUN instead.

**Comment by carriepl** [comment]
Travis has blessed the result. I think every comment has been addressed now.
